### PR TITLE
Bypass eligible caches for implicit search under GADT constraints

### LINKED ISSUE #13974: dotty doesn't seem to take the full context into account when resolving implicits
## Compiler version

3.1.0 (also tested on latest master `3.1.2-RC1-bin-SNAPSHOT-nonbootstrapped-git-2b37257`)

## Minimized code

```Scala
object Test {
  sealed abstract class Use[A]
  case class UseArray[A]() extends Use[Array[A]]

  def go[X](u: Use[X], x: X): Unit = {
    u match {
      case UseArray() =>
        x.size
    }
  }
}
```

## Output

```
-- [E008] Not Found Error: /home/bartosz/work/programming/scala/scalac/test/src/main/scala/Main.scala:8:10 
8 |        x.size
  |        ^^^^^^
  |        value size is not a member of X
  |
  |        where:    X is a type in method go which is an alias of Array[A$1]

```

## Expectation
Code compiles without errors.

### PR DESCRIPTION
**Pull request by odersky**
Under a non-empty GADT constraint, bypass caches when computing eligible implicits.
The GADT constraint might influence what implicits are available locally, so cached
results would be unreliable.

What this means for performance depends on how many implicit searches are performed
under non-empty GADT constraints. I measured for dotty itself. There, about 2% of givens
were checked in uncached mode because they were under a nonempty GADT constraint. 
Rechecking those 2% might have been avoided if they had been cached already otherwise.

Fixes #13974

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

**Review by abgruszecki — APPROVED** [review]
LGTM, we should merge.

**Comment by abgruszecki** [comment]
@odersky so what's the policy right now on merging PRs into master? I think this PR fits into a patch release, so I guess this means we can just merge it, right?

File: compiler/src/dotty/tools/dotc/core/GadtConstraint.scala [hunk 1]
```diff
@@ -304,7 +312,7 @@ final class ProperGadtConstraint private(
 
   override def fresh = new ProperGadtConstraint
   override def restore(other: GadtConstraint): Unit =
-    if (!other.isEmpty) sys.error("cannot restore a non-empty GADTMap")
+    assert(!other.isNarrowing, "cannot restore a non-empty GADTMap")
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/GadtConstraint.scala:315]:
```suggestion
    assert(other eq EmptyGadtConstraint, "cannot restore a non-empty GADTMap")
```

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/GadtConstraint.scala:315]: let's try this...

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/GadtConstraint.scala:315]: Did not work, reverting...

**Pull request merged**